crates/proto/src/udp/udp_stream.rs

```diff
@@ -25,18 +25,40 @@ use crate::Time;
 #[async_trait]
 pub trait UdpSocket
 where
-    Self: Sized + Unpin,
+    Self: Send + Sync + Sized + Unpin,
 {
     /// Time implementation used for this type
     type Time: Time;
 
     /// UdpSocket
     async fn bind(addr: &SocketAddr) -> io::Result<Self>;
+
+    /// Poll once Receive data from the socket and returns the number of bytes read and the address from
+    /// where the data came on success.
+    fn poll_recv_from(
+        &self,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<(usize, SocketAddr)>>;
+
     /// Receive data from the socket and returns the number of bytes read and the address from
     /// where the data came on success.
-    async fn recv_from(&mut self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)>;
+    async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
+        futures_util::future::poll_fn(|cx| self.poll_recv_from(cx, buf)).await
+    }
+
+    /// Poll once to send data to the given address.
+    fn poll_send_to(
+        &self,
+        cx: &mut Context,
+        buf: &[u8],
+        target: &SocketAddr,
```

Comment:
Just as an FYI, this was an error in the tokio api, it should be `SocketAddr`. There's been a PR to fix it, but it's a breaking API change so I don't know when they will actually release it. Might want to make a note about it.

https://github.com/tokio-rs/tokio/pull/3125

Reply:
Thanks for the feedback. I updated our interfaces to use the SocketAddr via Copy rather than pass-by-ref. When Tokio changes it, it will be easy to change without a breaking change in trust-dns.